libbeat/idxmgmt/ilm/eshandler.go

```diff
@@ -0,0 +1,210 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package ilm
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type esClientHandler struct {
+	client ESClient
+}
+
+var (
+	esMinILMVersion       = common.MustNewVersion("6.6.0")
+	esMinDefaultILMVesion = common.MustNewVersion("7.0.0")
+)
+
+const (
+	// esFeaturesPath is used to query Elasticsearch for availability of licensed
+	// features.
+	esFeaturesPath = "/_xpack"
+
+	esILMPath = "/_ilm/policy"
+
+	esAliasPath = "/_alias"
+)
+
+// ESClientHandler creates a new APIHandler executing ILM, and alias queries
+// against Elasticsearch.
+func ESClientHandler(client ESClient) APIHandler {
+	if client == nil {
+		return nil
+	}
+	return &esClientHandler{client}
+}
+
+// ESClient defines the minimal interface required for the ESClientHandler to
+// prepare a policy and write alias.
+type ESClient interface {
+	GetVersion() common.Version
+	Request(
+		method, path string,
+		pipeline string,
+		params map[string]string,
+		body interface{},
+	) (int, []byte, error)
+}
+
+func (h *esClientHandler) ILMEnabled(mode Mode) (bool, error) {
+	if mode == ModeDisabled {
+		return false, nil
+	}
+
+	avail, probe := h.checkILMVersion(mode)
+	if !avail {
+		if mode == ModeEnabled {
+			ver := h.client.GetVersion()
+			return false, errf(ErrESVersionNotSupported,
+				"Elasticsearch %v does not support ILM", ver.String())
+		}
+		return false, nil
+	}
+
+	if !probe {
+		// version potentially supports ILM, but mode + version indicates that we
+		// want to disable ILM support.
+		return false, nil
+	}
+
+	avail, enabled, err := h.checkILMSupport()
+	if err != nil {
+		return false, err
+	}
+
+	if !avail {
+		if mode == ModeEnabled {
+			return false, errOf(ErrESVersionNotSupported)
+		}
+		return false, nil
+	}
+
+	if !enabled && mode == ModeEnabled {
+		return false, errOf(ErrESILMDisabled)
+	}
+	return enabled, nil
+}
+
+func (h *esClientHandler) CreateILMPolicy(policy Policy) error {
+	path := path.Join(esILMPath, policy.Name)
+	_, _, err := h.client.Request("PUT", path, "", nil, policy.Body)
+	return err
+}
+
+func (h *esClientHandler) HasILMPolicy(name string) (bool, error) {
+	// XXX: HEAD method does currently not work for checking if a policy exists
+	path := path.Join(esILMPath, name)
+	status, b, err := h.client.Request("GET", path, "", nil, nil)
+	if err != nil && status != 404 {
+		return false, wrapErrf(err, ErrRequestFailed,
+			"failed to check for policy name '%v': (status=%v) %s", name, status, b)
+	}
+	return status == 200, nil
+}
+
+func (h *esClientHandler) HasAlias(name string) (bool, error) {
```

Comment:
I am looking forward to having an Elasticseach Golang client. I had to implement the same functions in the migration tool.